File: ra-idp-go/internal/oauth2/domain/pkce.go
// Package domain: OAuth2 ドメインモデル。
//
// PKCE (RFC 7636) verifier の検証。code_challenge_method=S256 のみサポート。
package domain

import (
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
)

// VerifyPKCES256 は code_verifier の SHA-256 → base64url(no-padding) が
// code_challenge と一致するかを検証する。
func VerifyPKCES256(codeVerifier, codeChallenge string) bool {
	sum := sha256.Sum256([]byte(codeVerifier))
	expected := base64.RawURLEncoding.EncodeToString(sum[:])
	return subtle.ConstantTimeCompare([]byte(expected), []byte(codeChallenge)) == 1
}

File: ra-oauth2-idp/src/domain/pkce.ts
/**
 * Layer 3 — Application Logic
 *
 * PKCE (RFC 7636) 検証ロジック。
 * 仕様核では「SHA-256 の S256 method」のみを許容する（ADR-002）。
 */

import { createHash, timingSafeEqual } from 'crypto'

/**
 * 与えられた code_verifier が code_challenge と一致するか検証する。
 * 一定時間比較でタイミング攻撃を防ぐ。
 */
export function verifyPkce(verifier: string, challenge: string, method: 'S256' = 'S256'): boolean {
  if (method !== 'S256') return false
  const computed = createHash('sha256').update(verifier).digest('base64url')
  if (computed.length !== challenge.length) return false
  return timingSafeEqual(Buffer.from(computed), Buffer.from(challenge))
}
